keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+Provisioners are expected to create a new bucket and related artifacts such as user, policies, credentials, etc.
+Provisioners return a skeleton OB structure.
+
+- **`Grant`** is a method called by the library when a new OBC is detected and its storage class contains the bucket name, meaning "brownfield" provisioning.
+In this case the OBC does not contain the bucket name.
+Provisioners are expected to create artifacts such as user, policies, credentials, etc., but not to create a new bucket.
+Provisioners return a skeleton OB structure.
+
+- **`Delete`** is a method called by the library when an OBC is deleted, and its storage class does not contain the bucket name (meaning "greenfield" provisioning had occurred), and the storage class's `reclaimPolicy` is "Delete".
+Provisioners are expected to remove the bucket and related artifacts.
+
+- **`Revoke`** is a method called by the library when an OBC is deleted and one of the following situations exists.
+  - the OBC's storage class contains the bucket name, meaning "brownfield" provisioning had occurred.
+In this case the storage class's `reclaimPolicy` is ignored.
+  - "greenfield" provisioning occurred and the storage class's `reclaimPolicy` is "Retain".
+In both cases provisioners are expected to retain the bucket but delete the object-store related artifacts.
+  
+### Binding
+
+Bucket binding refers to the steps necessary to make the target bucket available to a pod.
+For _greenfield_ a new bucket is physically created.
+In both new and existing bucket use cases Kubernetes resources are created by the library: a secret containing access credentials, a configMap containing bucket endpoint info, and an OB describing the bucket.
+Each provisioner will likely have to create and delete their own object store-specific resources.
+
+Bucket binding requires these steps before the bucket is accessible to an app pod:
+1. (greenfield only) generation of a random bucket name when requested (performed by bucket lib).
+1. (greenfield only) the creation of the physical bucket with owner credentials (performed by provisioner).
+1. creation of the necessary object store-specific resources, e.g. IAM user, policy, etc.
+1. creation of an OB, residing in the object-store's namespace, based on the provisioner's returned OB structure (performed by bucket lib).
+1. creation of a ConfigMap, residing in the OBC's namespace, based on the provisioner's returned endpoint info (performed by bucket lib).
+1. creation of a Secret, residing in the OBC's namespace, based on the provisioner's returned credentials (performed by bucket lib).
+
+### Custom Resource Details
+
+#### OBC Details (created by user)
+
+```yaml
+apiVersion: objectbucket.io/v1alpha1
+kind: ObjectBucketClaim
+metadata:
+  name: MY-BUCKET-1 [1]
+  namespace: USER-NAMESPACE [2]
+spec:
+  bucketName: [3]
+  generateBucketName: "photo-booth" [4]
+  storageClassName: AN-OBJECT-STORE-STORAGE-CLASS [5]
+  additionalConfig: [6]
```

Comment:
This feels like non-portable information, akin to passing through annotations.

Reply:
True. We want a way to pass provisioner specific information. The storage class (or bucketClass) allows this, but we were thinking of the case where, say region, was defaulted in the SC but could be overridden in the OBC. Of course each provisioner decides which of these fields to process or ignore.